source/parser/x86_64/allocators.hpp

```diff
@@ -53,35 +54,59 @@ namespace smeagle::x86_64 {
     }
 
     // Given two registers, return one combined string
-    std::string getRegistersString(std::pair<RegisterClass, RegisterClass> regClasses,
-                                   st::Type *paramType) {
-      std::string locA = this->getRegisterString(regClasses.first, paramType);
-      std::string locB = this->getRegisterString(regClasses.second, paramType);
-
-      // If B is empty (NO_CLASS) then return A
-      if (locB == "") {
-        return locA;
+    std::string getRegisterString(RegisterClass lo, RegisterClass hi, st::Type *paramType) {
+      if (lo == RegisterClass::NO_CLASS) {
+        throw std::runtime_error{"Can't allocate a {NO_CLASS, *}"};
+      }
+
+      if (lo == RegisterClass::MEMORY) {
+        // goes on the stack
+        return fallocator.nextFramebaseFromType(paramType);
       }
-      return locA + "|" + locB;
-    }
 
-    // Get a string location from a register class
-    std::string getRegisterString(RegisterClass regClass, st::Type *paramType) {
-      std::optional<std::string> regString;
+      if (lo == RegisterClass::INTEGER) {
+        auto reg = getNextIntRegister();
+        if (!reg) {
+          // Ran out of registers, put it on the stack
+          return fallocator.nextFramebaseFromType(paramType);
+        }
+        return reg.value();
+      }
 
-      // If the class is memory, pass the argument on the stack
-      if (regClass == RegisterClass::NO_CLASS) regString = "";
-      if (regClass == RegisterClass::SSE) regString = this->getNextSseRegister();
-      if (regClass == RegisterClass::INTEGER) regString = this->getNextIntRegister();
-      if (regClass == RegisterClass::MEMORY) regString = std::nullopt;
+      if (lo == RegisterClass::SSE) {
```

Comment:
Is there any reason to not have these as their own functions (like before?)

Reply:
The `getRegistersString` function wasn't correct because x86 doesn't allocate register space for both pieces of the classification. Rather, their characteristics are combined to allocate a single register. For example, an aggregate with two 32-bit `int`s would have the classification `{INTEGER,INTEGER}`, but would only get allocated a single register.